### PR TITLE
feat: add keyword argument support for `train_from_stream` (#114)

### LINKED ISSUE #114: feat: Add keyword argument support for train_from_stream
## Summary

The `train_from_stream` method currently only accepts positional arguments:

```ruby
classifier.train_from_stream(:spam, File.open("spam_corpus.txt"))
```

For API consistency with other methods like `train` and `add`, it would be nice to also support keyword arguments:

```ruby
classifier.train_from_stream(spam: File.open("spam_corpus.txt"))
```

## Affected Classes

- `Classifier::Bayes`
- `Classifier::LSI`
- `Classifier::KNN`
- `Classifier::LogisticRegression`

## Current Behavior

```ruby
classifier.train_from_stream(spam: File.open("corpus.txt"))
# => ArgumentError: wrong number of arguments (given 0, expected 2)
```

## Expected Behavior

```ruby
classifier.train_from_stream(spam: File.open("corpus.txt"))
# => Works, trains the :spam category from the stream
```

### PR DESCRIPTION
**Pull request by Yegorov**
Closes #114

### CODE REVIEW

File: lib/classifier/streaming.rb [hunk 1]
```diff
@@ -26,8 +26,8 @@ module Streaming
     # Trains the classifier from an IO stream.
     # Each line in the stream is treated as a separate document.
     #
-    # @rbs (Symbol | String, IO, ?batch_size: Integer) { (Progress) -> void } -> void
-    def train_from_stream(category, io, batch_size: DEFAULT_BATCH_SIZE, &block)
+    # @rbs (Symbol | String, IO, ?batch_size: Integer, **Hash[Symbol, IO]) { (Progress) -> void } -> void
+    def train_from_stream(category, io, batch_size: DEFAULT_BATCH_SIZE, **categories, &block)
```
[the remark this answers is not in the document]
Comment (Yegorov) [reply lib/classifier/streaming.rb]: @greptile-apps fixed!

File: lib/classifier/bayes.rb [hunk 2]
```diff
@@ -328,20 +328,23 @@ def remove_category(category)
     #     puts "#{progress.completed} documents processed"
     #   end
     #
-    # @rbs (String | Symbol, IO, ?batch_size: Integer) { (Streaming::Progress) -> void } -> void
-    def train_from_stream(category, io, batch_size: Streaming::DEFAULT_BATCH_SIZE)
-      category = category.prepare_category_name
-      raise StandardError, "No such category: #{category}" unless @categories.key?(category)
+    # @rbs (?(String | Symbol), ?IO, ?batch_size: Integer, **Hash[Symbol, IO]) { (Streaming::Progress) -> void } -> void
+    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories)
+      (category && io ? { category => io } : categories).each do |category, io|
```
[the remark this answers is not in the document]
Comment (Yegorov) [reply lib/classifier/bayes.rb]: @greptile-apps fixed!

File: lib/classifier/lsi.rb [hunk 3]
```diff
@@ -662,25 +662,29 @@ def self.load_checkpoint(storage:, checkpoint_id:)
...
+            yield progress if block_given?
+          end
+        ensure
+          @auto_rebuild = original_auto_rebuild
+          build_index if original_auto_rebuild
         end
-      ensure
-        @auto_rebuild = original_auto_rebuild
-        build_index if original_auto_rebuild
       end
```
[the remark this answers is not in the document]
Comment (Yegorov) [reply lib/classifier/lsi.rb]: @greptile-apps fixed!

File: lib/classifier/bayes.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yegorov) [reply lib/classifier/bayes.rb]: @greptile-apps ok, but how i can fix this rubocop offenses?
```
lib/classifier/bayes.rb:332:5: C: Metrics/CyclomaticComplexity: Cyclomatic complexity for train_from_stream is too high. [11/10]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/classifier/bayes.rb:332:5: C: Metrics/PerceivedComplexity: Perceived complexity for train_from_stream is too high. [11/10]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/classifier/logistic_regression.rb:394:5: C: Metrics/AbcSize: Assignment Branch Condition size for train_from_stream is too high. [<16, 30, 12> 36.06/30]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/classifier/logistic_regression.rb:394:5: C: Metrics/CyclomaticComplexity: Cyclomatic complexity for train_from_stream is too high. [13/10]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/classifier/logistic_regression.rb:394:5: C: Metrics/PerceivedComplexity: Perceived complexity for train_from_stream is too high. [13/10]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/classifier/lsi.rb:666:5: C: Metrics/CyclomaticComplexity: Cyclomatic complexity for train_from_stream is too high. [12/10]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/classifier/lsi.rb:666:5: C: Metrics/PerceivedComplexity: Perceived complexity for train_from_stream is too high. [12/10]
    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories) ...

```

[the remark this answers is not in the document]
Comment (Yegorov) [reply lib/classifier/bayes.rb]: @greptile-apps thanks, fixed!

**Comment by Yegorov** [comment]
@cardmagic can you take a look, please?

File: lib/classifier/lsi.rb [hunk 4]
```diff
@@ -662,26 +662,19 @@ def self.load_checkpoint(storage:, checkpoint_id:)
...
     #   end
     #
-    # @rbs (String | Symbol, IO, ?batch_size: Integer) { (Streaming::Progress) -> void } -> void
-    def train_from_stream(category, io, batch_size: Streaming::DEFAULT_BATCH_SIZE)
+    # @rbs (?(String | Symbol | nil), ?IO?, ?batch_size: Integer, **IO) { (Streaming::Progress) -> void } -> void
+    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories, &)
+      raise ArgumentError, 'Provide either (category, io) or keyword category: io pairs' if category.nil? && io.nil? && categories.empty?
+      raise ArgumentError, 'Provide both category and io, or use keyword arguments' if category.nil? ^ io.nil?
+
       original_auto_rebuild = @auto_rebuild
```
Comment (cardmagic) [inline lib/classifier/lsi.rb]: **Must-fix #1 — `ensure` clobbers `@auto_rebuild` to `nil` when validation raises.**

The two `raise ArgumentError` on lines 667-668 execute **before** `original_auto_rebuild = @auto_rebuild` on line 670. But the method-level `ensure` (line 675) always runs. Ruby has already declared `original_auto_rebuild` as a local in scope, so it's `nil` in the ensure block — meaning line 676 sets `@auto_rebuild = nil`, silently breaking subsequent operations.

Reproduced against a real `Classifier::LSI` instance:
```
Before: auto_rebuild=true
Caught ArgumentError
After:  auto_rebuild=nil   ← clobbered
```

KNN inherits this since it delegates to LSI.

**Fix:** hoist `original_auto_rebuild = @auto_rebuild` above the two validation raises, or wrap the body in an explicit `begin/ensure` that starts after validation.

File: lib/classifier/lsi.rb [hunk 5]
```diff
@@ -729,6 +722,23 @@ def train_batch(category = nil, documents = nil, batch_size: Streaming::DEFAULT_
 
     private
 
+    # Trains from an IO stream with a single category.
+    # @rbs (String | Symbol, IO, batch_size: Integer) { (Streaming::Progress) -> void } -> void
+    def stream_train_category(category, io, batch_size:)
+      raise StandardError, 'Stream must respond to #each_line' unless io.respond_to?(:each_line)
```
Comment (cardmagic) [inline lib/classifier/lsi.rb]: **Must-fix #2 — IO validation happens mid-loop after partial mutation.**

When called as `train_from_stream(a: good_io, b: Object.new)`, the first stream's documents are added to the index before the second iteration raises. State is now partially mutated.

**Fix:** pre-validate all IOs in one pass before the `.each` loop in `train_from_stream`, so the call either fully succeeds or raises before touching state.

File: lib/classifier/streaming.rb [hunk 6]
```diff
@@ -26,8 +26,8 @@ module Streaming
     # Trains the classifier from an IO stream.
     # Each line in the stream is treated as a separate document.
     #
-    # @rbs (Symbol | String, IO, ?batch_size: Integer) { (Progress) -> void } -> void
-    def train_from_stream(category, io, batch_size: DEFAULT_BATCH_SIZE, &block)
+    # @rbs (?(Symbol | String | nil), ?IO?, ?batch_size: Integer, **Hash[Symbol, IO]) { (Progress) -> void } -> void
```
Comment (cardmagic) [inline lib/classifier/streaming.rb]: **Must-fix #3 — base-module type signature mismatches the implementations.**

In RBS, `**T` is the type of each rest-kwarg value. So:
- `**IO` (all four implementations) = each kwarg value is an `IO` — matches the runtime: `train_from_stream(spam: io1)` passes an IO as the value
- `**Hash[Symbol, IO]` (this base stub) = each kwarg value is itself a `Hash[Symbol, IO]` — would mean `train_from_stream(spam: { foo: io })`

The base stub should match:

```ruby
# @rbs (?(Symbol | String | nil), ?IO?, ?batch_size: Integer, **IO) { (Progress) -> void } -> void
```

This may also let you drop the `# @type var categories: untyped` escape hatch in `knn.rb:273`.

File: lib/classifier/bayes.rb [hunk 7]
```diff
@@ -389,6 +382,25 @@ def self.load_checkpoint(storage:, checkpoint_id:)
 
     private
 
+    # Trains from an IO stream with a single category.
+    # @rbs (String | Symbol, IO, batch_size: Integer) { (Streaming::Progress) -> void } -> void
+    def stream_train_category(category, io, batch_size:)
+      category = category.prepare_category_name
+      raise StandardError, "No such category: #{category}" unless @categories.key?(category)
+      raise StandardError, 'Stream must respond to #each_line' unless io.respond_to?(:each_line)
```
Comment (cardmagic) [inline lib/classifier/bayes.rb]: **Should-fix #4 — use `ArgumentError`, not `StandardError`.**

`StandardError` is the base of nearly every Ruby exception; rescuers can't distinguish bad-input from genuine errors. The public `train_from_stream` correctly raises `ArgumentError` (line 333-334); the private helper should too. Same issue in `logistic_regression.rb:432-433` and `lsi.rb:728`.

The existing `assert_raises(StandardError)` tests still pass since `ArgumentError < StandardError`, but should be tightened to `assert_raises(ArgumentError)`.

File: lib/classifier/bayes.rb [hunk 8]
```diff
@@ -328,20 +328,13 @@ def remove_category(category)
...
-      reader = Streaming::LineReader.new(io, batch_size: batch_size)
-      total = reader.estimate_line_count
-      progress = Streaming::Progress.new(total: total)
-
-      reader.each_batch do |batch|
-        train_batch_internal(category, batch)
-        progress.completed += batch.size
-        progress.current_batch += 1
-        yield progress if block_given?
+      (category && io ? { category => io } : categories).each do |(category, io)|
```
Comment (cardmagic) [inline lib/classifier/bayes.rb]: **Nit — unnecessary destructuring + reused variable names.**

Two small things in `(category && io ? { category => io } : categories).each do |(category, io)|`:

1. The block params reuse the method-parameter names `category` and `io`; inside the block the outer values are inaccessible.
2. The `|(...)|` destructuring is unnecessary — `Hash#each` already yields two args.

Cleaner:

```ruby
pairs = category ? { category => io } : categories
pairs.each do |cat, stream|
  stream_train_category(cat, stream, batch_size: batch_size, &)
end
```

Same pattern at `lsi.rb:672` and `logistic_regression.rb:398`.

File: lib/classifier/bayes.rb [hunk 9]
```diff
@@ -328,20 +328,13 @@ def remove_category(category)
...
     #   end
     #
-    # @rbs (String | Symbol, IO, ?batch_size: Integer) { (Streaming::Progress) -> void } -> void
-    def train_from_stream(category, io, batch_size: Streaming::DEFAULT_BATCH_SIZE)
-      category = category.prepare_category_name
-      raise StandardError, "No such category: #{category}" unless @categories.key?(category)
+    # @rbs (?(String | Symbol | nil), ?IO?, ?batch_size: Integer, **IO) { (Streaming::Progress) -> void } -> void
+    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories, &)
+      raise ArgumentError, 'Provide either (category, io) or keyword category: io pairs' if category.nil? && io.nil? && categories.empty?
+      raise ArgumentError, 'Provide both category and io, or use keyword arguments' if category.nil? ^ io.nil?
```
Comment (cardmagic) [inline lib/classifier/bayes.rb]: **Nit — XOR on booleans reads as Ruby exotica.**

`category.nil? ^ io.nil?` works but `^` on booleans is rare enough in idiomatic Ruby to make readers pause. Consider:

```ruby
raise ArgumentError, '...' if category.nil? != io.nil?
```

or

```ruby
raise ArgumentError, '...' if [category, io].one?(&:nil?)
```

File: lib/classifier/knn.rb [hunk 10]
```diff
@@ -268,9 +268,10 @@ def self.load_checkpoint(storage:, checkpoint_id:)
     #     puts "#{progress.completed} documents processed"
     #   end
     #
-    # @rbs (String | Symbol, IO, ?batch_size: Integer) { (Streaming::Progress) -> void } -> void
-    def train_from_stream(category, io, batch_size: Streaming::DEFAULT_BATCH_SIZE, &block)
-      @lsi.train_from_stream(category, io, batch_size: batch_size, &block)
+    # @rbs (?(String | Symbol | nil), ?IO?, ?batch_size: Integer, **IO) { (Streaming::Progress) -> void } -> void
+    def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories, &block)
```
Comment (cardmagic) [inline lib/classifier/knn.rb]: **Nit — inconsistent block forwarding.**

KNN uses `&block` here but never references `block` by name; the other three classifiers all use anonymous `&` (Ruby 3.1+). Switch for consistency:

```ruby
def train_from_stream(category = nil, io = nil, batch_size: Streaming::DEFAULT_BATCH_SIZE, **categories, &)
  @lsi.train_from_stream(category, io, batch_size: batch_size, **categories, &)
  synchronize { @dirty = true }
end
```

Also: when must-fix #3 is fixed, the `# @type var categories: untyped` on line 273 may no longer be needed.

File: lib/classifier/logistic_regression.rb [hunk 11]
```diff
@@ -440,6 +425,33 @@ def train_batch(category = nil, documents = nil, batch_size: Streaming::DEFAULT_
 
     private
 
+    # Trains from an IO stream with a single category.
+    # @rbs (String | Symbol, IO, batch_size: Integer) { (Streaming::Progress) -> void } -> void
+    def stream_train_category(category, io, batch_size:)
+      category = category.to_s.prepare_category_name
+      raise StandardError, "No such category: #{category}" unless @categories.include?(category)
+      raise StandardError, 'Stream must respond to #each_line' unless io.respond_to?(:each_line)
```
Comment (cardmagic) [inline lib/classifier/logistic_regression.rb]: **Should-fix #4 (cross-ref) — `ArgumentError`, not `StandardError`.** See the comment on `bayes.rb:389-390`.

File: test/bayes/streaming_test.rb [hunk 12]
```diff
@@ -17,6 +17,23 @@ def test_train_from_stream_basic
...
+
+    assert_equal 'Spam', classifier.classify('buy free')
+    assert_equal 'Ham', classifier.classify('hello meeting')
+  end
+
+  def test_train_from_stream_invalid_io_type
+    assert_raises(StandardError) do
+      @classifier.train_from_stream(spam: Object.new)
+    end
+  end
```
Comment (cardmagic) [inline test/bayes/streaming_test.rb:35]: **Test coverage gap.**

The new tests don't exercise the public `ArgumentError` validation paths (`train_from_stream()` with no args, or `train_from_stream(:spam)` without `io`). No test would have caught must-fix #1. Suggested additions:

```ruby
def test_train_from_stream_raises_without_args
  assert_raises(ArgumentError) { @classifier.train_from_stream }
end

def test_train_from_stream_raises_with_partial_args
  assert_raises(ArgumentError) { @classifier.train_from_stream(:spam) }
end
```

LSI additionally needs a test for `auto_rebuild: true` combined with the multi-category keyword form (the existing `test_train_from_stream_rebuilds_index_when_auto_rebuild` only uses two positional calls). And a test for heterogeneous valid+invalid kwargs would lock in must-fix #2.

**Review by cardmagic — COMMENTED** [review]
The keyword-arg API is a nice addition and the multi-category path works — issues are around the edges (validation ordering, state mutation, base-module rbs). Test suite passes locally (678 runs, 0 failures).

Note: Greptile's earlier review is stale — it flagged a "build_index per-category" regression and "silent no-op for partial args" that you already fixed in a later commit. Those no longer apply.

See inline comments. Must-fix: #1, #2, #3. Should-fix: #4. The rest are nits/suggestions.

File: lib/classifier/lsi.rb (hunk 4 above)
Comment (Yegorov) [reply lib/classifier/lsi.rb]: @cardmagic thanks, fixed!

File: lib/classifier/lsi.rb (hunk 5 above)
Comment (Yegorov) [reply lib/classifier/lsi.rb]: @cardmagic thanks, fixed!

File: lib/classifier/streaming.rb (hunk 6 above)
Comment (Yegorov) [reply lib/classifier/streaming.rb]: @cardmagic thanks, fixed! 
> This may also let you drop the `# @type var categories: untyped` escape hatch in knn.rb:273.

But `# @type var categories: untyped` comment it's still needed to avoid errors.

File: lib/classifier/bayes.rb (hunk 7 above)
Comment (Yegorov) [reply lib/classifier/bayes.rb]: @cardmagic thanks, fixed!

File: lib/classifier/bayes.rb (hunk 8 above)
Comment (Yegorov) [reply lib/classifier/bayes.rb]: @cardmagic thanks, fixed!

File: lib/classifier/bayes.rb (hunk 9 above)
Comment (Yegorov) [reply lib/classifier/bayes.rb]: @cardmagic thanks, fixed!

File: lib/classifier/knn.rb (hunk 10 above)
Comment (Yegorov) [reply lib/classifier/knn.rb]: @cardmagic thanks, fixed!
> Also: when must-fix #3 is fixed, the `# @type var categories: untyped` on line 273 may no longer be needed.

Unfortunately, this has no effect.

File: lib/classifier/logistic_regression.rb (hunk 11 above)
Comment (Yegorov) [reply lib/classifier/logistic_regression.rb]: @cardmagic thanks, fixed!

File: test/bayes/streaming_test.rb (hunk 12 above)
Comment (Yegorov) [reply test/bayes/streaming_test.rb:35]: @cardmagic thanks, fixed!

**Review by cardmagic — APPROVED** [review]

**Comment by cardmagic** [comment]
Hi @Yegorov,

Thanks for the great work on PR #155. The multi-category `train_from_stream` API is a really nice addition. A few things I appreciated:

- **Comprehensive coverage**: adding the keyword form to all four classifiers (Bayes, KNN, LSI, LogisticRegression) rather than just one
- **Solid validation**: the explicit `ArgumentError` guards for partial/missing args and invalid IO types are exactly the right call
- **Strong tests**: including the auto_rebuild edge cases for LSI
- **Responsive to review**: turning around the Greptile feedback quickly and cleanly, especially the LSI `build_index` hoisting fix

This shipped as classifier 2.5.0 on RubyGems. Thanks also for closing out #114, and for your continued contributions to the gem.

Lucas
